### ISSUE TITLE
update-aurto fails silently due to repo desync

### ISSUE THREAD

**Issue by ScoreUnder** [alexheretic/aurto#22]
To preface, I am a strong opponent of the use of `set -e`, but I shall try not to let that get the better of me here.

The update-aurto script currently fails silently for me:
```
% sudo systemctl status update-aurto
● update-aurto.service - Updates 'aurto' repo                  
   Loaded: loaded (/usr/lib/systemd/system/update-aurto.service; static; vendor preset: disabled)
   Active: failed (Result: exit-code) since Mon 2019-07-29 01:12:55 BST; 4min 26s ago
  Process: 53039 ExecStart=/usr/lib/aurto/update-aurto (code=exited, status=1/FAILURE)
 Main PID: 53039 (code=exited, status=1/FAILURE)

Jul 29 01:12:55 kirisame systemd[1]: Started Updates 'aurto' repo.
Jul 29 01:12:55 kirisame update-aurto[53039]: aurto: Checking maintainer trust...
Jul 29 01:12:55 kirisame sudo[53189]:     root : TTY=unknown ; PWD=/ ; USER=score ; COMMAND=/usr/bin/aurto remove qt4
Jul 29 01:12:55 kirisame sudo[53189]: pam_unix(sudo:session): session opened for user score by (uid=0)
Jul 29 01:12:55 kirisame sudo[53189]: pam_unix(sudo:session): session closed for user score
Jul 29 01:12:55 kirisame systemd[1]: update-aurto.service: Main process exited, code=exited, status=1/FAILURE
Jul 29 01:12:55 kirisame systemd[1]: update-aurto.service: Failed with result 'exit-code'.
```

After running it under `bash -x`, it says this:
```
% sudo bash -x /usr/lib/aurto/update-aurto
+ set -eu
+ readonly 'AURVCS=.*-(bzr|git|hg|svn)$'
+ AURVCS='.*-(bzr|git|hg|svn)$'
+ [[ 0 -ne 0 ]]
+++ dirname /usr/lib/aurto/update-aurto
++ cd /usr/lib/aurto
++ pwd
+ lib_dir=/usr/lib/aurto
++ cat /usr/lib/aurto/user
+ user=score
+ source /usr/lib/aurto/shared-functions
++ test -t 1
++ export chroot_arg=--chroot
++ chroot_arg=--chroot
++ test -f /usr/lib/aurto/conf-disable-chroot
+ '[' -z score ']'
+ pacsync aurto
+ '[' -f /etc/aurto/trusted-users ']'
+ echo 'aurto: Checking maintainer trust...'
aurto: Checking maintainer trust...
+++ pacman -Slq aurto
++ /usr/lib/aurto/trust-check {snip huge package list}
+ mistrust=qt4:dviktor
+ '[' -z qt4:dviktor ']'
++ not_in_aur_packages qt4:dviktor
++ local packages=
++ read -r line
++ '[' -n qt4:dviktor ']'
++ [[ qt4:dviktor = *\:\:\n\o\t\-\i\n\-\a\u\r* ]]
++ read -r line
++ echo ''
++ sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//'
++ tr -s ' '
+ not_in_aur=
+++ echo qt4:dviktor
+++ grep -v ::
+++ cut -d: -f1
++ new_line_to_space_separated_unique qt4
++ local space_sep=
++ read -r line
++ '[' -n qt4 ']'
++ [[ '' != *\ \q\t\4\ * ]]
++ space_sep=' qt4 '
++ read -r line
++ echo ' qt4 '
++ sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//'
++ tr -s ' '
+ mistrusted_pkgs=qt4
+ '[' -n '' ']'
+ '[' -n qt4 ']'
+ aurto remove qt4
```

This is failing silently with `set -e`, but what's weird is that `aurto remove` is also failing silently.
```
% bash -x aurto remove qt4                  
+ set -eu
+ version=0.9.4
+ command=remove
+ arg1=qt4
+ [[ 1000 -eq 0 ]]
+++ dirname /bin/aurto
++ cd /bin
++ pwd
+ bin_dir=/bin
+ lib_dir=/bin/../lib/aurto
+ source /bin/../lib/aurto/shared-functions
++ test -t 1
++ export chroot_arg=--chroot
++ chroot_arg=--chroot
++ test -f /usr/lib/aurto/conf-disable-chroot
+ '[' remove == add ']'
+ '[' remove == addpkg ']'
+ '[' remove == remove ']'
+ '[' -n qt4 ']'
+ removed=
+ for pkg in "${@:2}"
++ repo-remove /var/cache/pacman/aurto/aurto.db.tar qt4
+ remove_out='==> Extracting database to a temporary location...
==> Extracting database to a temporary location...
==> Searching for package '\''qt4'\''...
==> ERROR: Package matching '\''qt4'\'' not found.
==> No packages modified, nothing to do.'
```

This seems to be a similar culprit: unexpected false exit code under `set -e`.

Not sure how the desynchronisation happened but the fix seemed to be to run:
```
sudo rm /var/lib/pacman/sync/aurto.db
sudo pacman -Syu
```
Which got the aurto database back in line with the pacman sync database (note that the rm was required; otherwise it falsely reported that it was up-to-date).
In the past I have had to poke around in the db before to remove things like delta files for packages that no longer exist, but otherwise I'm not sure what the cause might have been. This seems to have been more of a recent development, unrelated to that.

So I wanted to report this as two problems:
1. The db getting out of sync(?) causing failure in an early stage of update-aurto (maybe there is a cheap way to sync it before the script sets off properly?)
2. No error message for failed package removal

**Comment by alexheretic** [comment]
Thanks for the report. It seems that repo-remove is returning non-zero exit code when it can't find the package to remove, this seems correct so I'll just assume it didn't always do this before.

`aurto remove` shouldn't return non-zero this is now fixed in e73ff7fb8bcd500a3f53948875048f7ca267ccdf. 

update-aurto _does_ call `pacsync aurto` so the `pacman -Slq aurto` call should be up to date.

I wonder if this happens because you don't trust qt4:dviktor, qt4 isn't in the aurto repo, but qt4 _is_ a dependency of one the packages that is in the aurto repo. Could that be true?

This could mean that update-aurto should removing all mistrusted/missing packages **and** all packages that depended on them. Currently it looks like we only do the first.

**Comment by ScoreUnder** [comment]
I think your diagnosis is about right. I found qt4 lying around on my system but not actually in the aurto repo. There is a distinct possibility that I actually had it installed from back when it was in the arch repos, but I don't remember.

0.9.5 has fixed this issue for me, so closing. Thanks!
